File: 1309-decrypt-string-from-alphabet-to-integer-mapping/1309-decrypt-string-from-alphabet-to-integer-mapping.py
class Solution:
    def freqAlphabets(self, s: str) -> str:
        nums = [str(i) for i in range (1, 27)]
        letters = [let for let in 'abcdefghijklmonpqrstuvwxyz']
        mapx = dict(zip(nums, letters))
        strx=''
        stack=[]

        for chr in s:
            if chr != '#':
                stack.append(chr)
            else:
                strx += stack.pop(-1)
                strx += stack.pop(-1)
                stack.append(strx[::-1])
                strx=''

        print(stack)
        return ''.join([mapx[num] for num in stack])


        

File: 1309-decrypt-string-from-alphabet-to-integer-mapping/1309-decrypt-string-from-alphabet-to-integer-mapping.cpp
class Solution {
public:
    string freqAlphabets(string s) {
        unordered_map<string, char> mapx;
        for (int i = 1; i <= 26; ++i) {
            mapx[to_string(i)] = 'a' + (i - 1);
        }
        vector<string> stack;
        string strx;
        
        for (char chr : s) {
            if (chr != '#') {
                stack.push_back(string(1, chr));
            } else {
                strx += stack.back(); stack.pop_back();
                strx += stack.back(); stack.pop_back();
                reverse(strx.begin(), strx.end());
                stack.push_back(strx);
                strx.clear();
            }
        }

        //reuse strx which should be blank string instead of creting 'string result'
        for (const auto& num : stack) {
            strx += mapx[num];
        }
        return strx;
    }
};

